services/comments.js

```diff
@@ -49,70 +53,77 @@ module.exports = class CommentsService {
    * @throws if tag name is not in ALLOWED_TAGS
    * @param {String} id the id of the comment to tag
    * @param {String} name the name of the tag to add
-   * @param {String} assigned_by the user id for the user who added the tag
+   * @param {String} added_by the user id for the user who added the tag
    */
-  static addTag(id, name, assigned_by) {
-
-    // Disabling allowed tags until we are able to extend them
-    // if (ALLOWED_TAGS.find((t) => t.name === name) == null) {
-    //   return Promise.reject(new Error('tag not allowed'));
-    // }
-
-    const filter = {
-      id,
-      'tags.name': {$ne: name},
-    };
-    const update = {
-      $push: {tags: {
-        name,
-        assigned_by,
-        created_at: new Date()
-      }}
-    };
-    return CommentModel.update(filter, update)
-      .then(({nModified}) => {
-        switch (nModified) {
-        case 0:
-
-          // either the tag was already there, or the comment doesn't exist with that id...
-          throw new Error('Could not add tag to comment. Either the comment doesn\'t exist or the tag is already present.');
-        case 1:
-
-            // tag added
-          return;
-        default:
-
-          // this should never happen because no multi parameter and unique index on id
+  static addTag(id, name, added_by) {
+
+    // Check that the tag is allowed by the system OR in our setting.tags.
+    return SettingsService.retrieve()
+    .then((settings) => {
+
+      UsersService.findById(added_by)
+      .then((user) => {
+
+        // Moderators or ADMIN can add any tag automatically.
+        if (user != null && (user.hasRoles('ADMIN') || user.hasRoles('MODERATOR'))) {
+          SettingModel.findOneAndUpdate({id: settings.id}, {
+            $push: {
+              tags: {
+                id: name,
+                models: ['COMMENTS']
+              }
+            }
+          });
+        }
+        else if (!ALLOWED_TAGS.includes(name) || settings.tags.findIndex((t) => {return t.id === name & t.models.include('COMMENTS');}) === -1) {
+          return Promise.reject(errors.ErrTagNotAllowed);
         }
       });
+
+      return CommentModel.findOneAndUpdate({id, 'tags.id': {$ne: name}}, {
+        $push: {
+          tags: {
+            id: name,
+            added_by: added_by
+          }
+        },
+      })
+        .then(({nModified}) => {
+          switch (nModified) {
+          case 0:
+            return Promise.reject(errors.ErrNoCommentFound);
+          case 1:
+            return;
+          default:
+          }
+        });
+    });
```

Comment:
There's a few issues here, I'll push some changes that resolve some issues with promises + query paths.